io.openems.edge.evcs.spelsberg.smart/src/io/openems/edge/evcs/spelsberg/smart/EvcsSpelsbergSmartImpl.java

```diff
@@ -0,0 +1,323 @@
+		return this.currentToPower(this.config.minHwCurrent());
+	}
+
+	@Override
+	public int getConfiguredMaximumHardwarePower() {
+		return this.currentToPower(this.config.maxHwCurrent());
+	}
+
+	@Override
+	public boolean getConfiguredDebugMode() {
+		return this.config.debugMode();
+	}
+
+	@Override
+	public boolean applyChargePowerLimit(int power) throws Exception {
+		setApplyChargePowerLimit(power);
+		return true;
+	}
+
+	@Override
+	public boolean pauseChargeProcess() throws Exception {
+		this.applyChargePowerLimit(0);
+		return true;
+	}
+
+	@Override
+	public boolean applyDisplayText(String text) throws OpenemsException {
+		return false;
+	}
+
+	@Override
+	public int getMinimumTimeTillChargingLimitTaken() {
+		return 30;
+	}
+
+	@Override
+	public ChargeStateHandler getChargeStateHandler() {
+		return this.chargeStateHandler;
+	}
+
+	@Override
+	public void handleEvent(Event event) {
+		if (!this.isEnabled()) {
+			return;
+		}
+		switch (event.getTopic()) {
+		case EdgeEventConstants.TOPIC_CYCLE_EXECUTE_WRITE:
+			this.writeHandler.run();
+			break;
+		}
+	}
+
+	private void applyConfig(Config config) {
+		this.config = config;
+		this._setChargingType(ChargingType.AC);
+		this._setPhases(Phases.THREE_PHASE.getValue());
+		this._setFixedMinimumHardwarePower(this.currentToPower(config.minHwCurrent()));
+		this._setFixedMaximumHardwarePower(this.currentToPower(config.maxHwCurrent()));
+		this._setPowerPrecision(DEFAULT_POWER_RECISION); // 1A steps
+	}
+
+	private Integer currentToPower(Integer current) {
+		return Math.round(current / 1000f) * DEFAULT_VOLTAGE * getPhasesAsInt();
+	}
+
+	private void addStatusCallback() {
+		this.channel(EvcsSpelsbergSmart.ChannelId.CHARGE_POINT_STATE).onSetNextValue(s -> {
+			ChargePointState state = s.asEnum();
+
+			/**
+			 * Maps the EVCS state to a {@link Status}.
+			 */
+			switch (state) {
+			case CHARGING -> this._setStatus(Status.CHARGING);
+			case NO_PERMISSION, CHARGING_STATION_RESERVED -> this._setStatus(Status.CHARGING_REJECTED);
+			case ERROR -> this._setStatus(Status.ERROR);
+			case NO_VEHICLE_ATTACHED -> this._setStatus(Status.NOT_READY_FOR_CHARGING);
+			case CHARGING_PAUSED -> this._setStatus(Status.CHARGING_REJECTED);
+			case UNDEFINED -> this._setStatus(Status.UNDEFINED);
+			default -> this._setStatus(Status.UNDEFINED);
+			}
+			;
+
+			if (getLifeBit() == -1) {
+				this._setModbusCommunicationFailed(true);
+			} else {
+				this._setModbusCommunicationFailed(false);
+			}
+		});
+	}
+
+	/*
+	 * Handle automatic phase shift and reset the fixed hardware power limits
+	 */
+	private void addPhaseDetectionCallback() {
+		final Consumer<Value<Integer>> setPhasesCallback = ignore -> {
+
+			var phases = 0;
+			if (getChargePowerL1() > 0) {
+				phases++;
+			}
+			if (getChargePowerL2() > 0) {
+				phases++;
+			}
+			if (getChargePowerL3() > 0) {
+				phases++;
+			}
+
+			this._setPhases(phases);
+
+			// Set full power range from 1,38kW to 11kW
```

Comment:
@sebastianasen: I have a question regarding the fixedMinimumHardwarePower. 
Our chargingStation has a build-in feature to switch automatically between 1-phase and 3-phase mode. The charging power can vary between 1,38kW and 11kW and system will automatically switch phases based on the APPLY_CHARGE_POWER_LIMIT from openEMS. Switching will be around the threshold of 4,2kW (with delay and hysterese)

However is is not yet available for the modbus interface. This will be integrated in the next firmware update. 

The Question is: Is there any issue expected if the EMS will apply a charge power of e.g 2,5kW and the chargingStation will stay in 3-phase mode and load with a minimum of 4,2kW (3x6A)?

If there might be an issue we should consider changing this implemented behavior. Or can we leave it and add the new feature (automatic phase switch for modbus) being compatible with the current firmware without this option.

Reply:
This is very good to know.
Currently there is no workaround for switching the phases during a charge process, but it is on our (Fenecon) timeline to handle such situations.

I assume that the UI and some parts of the edge need to be changed if we want to use and show it correctly.
For now the Evcs WriteHandler uses a fitIntoMinMax Method, that would not a allow a setpoint below the minimum and maximum hardware limit and set 0 (Charge Pause) instead.

Reply:
Okay. So should the implementation regarding setting minimum and maximum hardware limit be changed?
Should we implement static 3 phase limits (4,2kW to 11kW) or is a full range 1,38 kW- 11kW supposed to work untill openEMS has updated and our new firmware is available?